Add reference for Python Test Framework (#18)
codewars/docs: content/codewars-python-test-framework-v2.md

# Python test framework for Codewars (v2)
(upgrade by [@Bubbler-4](https://github.com/Bubbler-4), Feb 2019)

<br>
<br>

You may find here the detailed information about the [test framework](https://github.com/codewars/python-test-framework/blob/master/codewars_test/test_framework.py) designed for CodeWars. The present version is extending the first one, so old katas are still compatible, but the use of the v2 is required for newly created katas or translations.

_Note: the test module is aliased with `test`, so you can call every tool provided with it using either `Test.method` or `test.method`._

<br>
<br>

---

## Overview

The test framework allowes to write named groups of tests holding other named groups of tests, containing themselves assertions.

* Assertions (using the methods detailed below) can actually be used at any level in the groups, or even outise of them (but you're advised to avoid that).
* Note that the code executed inside on group isn't stopped at the first failed assertion (unlees you use `allow_raise`. See below), so the full test suite is executed each time.

The basic setup for the tests follow this example:

```python

> @Test.describe('Fixed Tests')

Reviewer: Can you change to `@test`?

We'll continue to support `Test` for now, but in the future, I'd like to let users just import the test framework normally like `import codewars_test as test` (removing the runner magic). So I want the documentation to constantly use `test`.

Author: argh... I just did the complete opposite x)

Ok.

Author: done (the markdowns too)

## Tests grouping (blocks)

The groups are created using the following decorators:

```python
@Test.describe(test_name)
def _():

    @Test.it(test_case_name)
    def _():
        ...
         
    @Test.describe(subgroup_name)
    def _():
      ...

        @Test.it(test_case_subgroup_name)
        def _():
            ...
```

Those decorators automatically run the decorated function to launch the tests. Both print elapsed time of the block at the end, and `describe`s can be nested.

* `describe` blocks:

Those blocks always stay depolyed whatever the results of the tests are, menaing that if you put assertions directly in them, they will be visible in the output, as well as any other `it` block(s) it could contain. They are the basic block inwhich you can insert `it` blocks or other nested `describe` blocks if you need to create subdivisions in your test suite.

* `it` blocks:

On the contrary of the `describe` blocks, the `it` blocks are deployed in the output panel only when they contain failed assertions. Oterwise they are shrinked and only the name of the block and the number of assertions are visible.

Note that you _cannot nest_ the `it` blocks.

<br>
<br>

---

> # Tests interruption

Reviewer: Only use `h1` (`# heading`) once for the title of the document.

Author: yep, I made the correction.

Reviewer: ? I meant once at the top `# Python test framework for Codewars (v2)`. There are still this one and `# Available assertion methods`. Those should be `##` and any children increased by one level.

Author: mmh? I thought I already corrected that one too. I'll check

Some of the functions below can accept a named argument `allow_raise=False`.

If you change its value to `True`, the tests contained inside the current block will be interrupted at the first failed test. The executions are then going back to the parent block if it exists and the next part is executed.

> On some computation-wise intensive katas, this may be a good idea to use this feature so that the user doesn't have to wait a long time before getting feedback (or possibly befor etiming out and in that cas, he might never get ny feedback at all, which may be cumbersome).

Reviewer: ```suggestion
On some computation-wise intensive katas, this may be a good idea to use this feature so that the user doesn't have to wait a long time before getting feedback (or possibly before timing out and in that case, they might never get any feedback at all, which may be cumbersome).
```
